### ISSUE TITLE
Error disabling Shared VPC Host - Destroy

### ISSUE THREAD

**Issue by pfilourenco** [terraform-providers/terraform-provider-google#6177]
Hi

I'm trying to disable/destroy  google_compute_shared_vpc_host_project but looks like is giving error: 

"Error: Error disabling Shared VPC Host "my_project": googleapi: Error 400: Cannot disable project as a shared VPC host because it has active service projects., badRequest"

Before this I successfully destroyed google_compute_shared_vpc_service_project.

Any idea?

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Affected Resource(s)

* google_XXXXX

### Terraform Configuration Files

```tf
# Copy-paste your Terraform configurations here.
#
# For large Terraform configs, please use a service like Dropbox and share a link to the ZIP file.
# For security, you can also encrypt the files using our GPG public key:
#    https://www.hashicorp.com/security
#
# If reproducing the bug involves modifying the config file (e.g., apply a config,
# change a value, apply the config again, see the bug), then please include both:
# * the version of the config before the change, and
# * the version of the config after the change.
```

### Steps to Reproduce

1. `terraform apply`

### References

* #0000

**Comment by venkykuberan** [comment]
I can't repro the issue, trying multiple times. Do you see similar plan output ?, verify right project ids are shown in plan output.

```
# google_compute_shared_vpc_host_project.host will be destroyed
  - resource "google_compute_shared_vpc_host_project" "host" {
      - id      = "test-project1-274319" -> null
      - project = "test-project1-274319" -> null
    }

  # google_compute_shared_vpc_service_project.service1 will be destroyed
  - resource "google_compute_shared_vpc_service_project" "service1" {
      - host_project    = "test-project1-274319" -> null
      - id              = "test-project1-274319/zinc-authority-275122" -> null
      - service_project = "zinc-authority-275122" -> null
    }

Plan: 0 to add, 0 to change, 2 to destroy.

Do you really want to destroy all resources?
  Terraform will destroy all your managed infrastructure, as shown above.
  There is no undo. Only 'yes' will be accepted to confirm.

  Enter a value: yes

google_compute_shared_vpc_service_project.service1: Destroying... [id=test-project1-274319/zinc-authority-275122]
google_compute_shared_vpc_service_project.service1: Still destroying... [id=test-project1-274319/zinc-authority-275122, 10s elapsed]
google_compute_shared_vpc_service_project.service1: Destruction complete after 14s
google_compute_shared_vpc_host_project.host: Destroying... [id=test-project1-274319]
google_compute_shared_vpc_host_project.host: Still destroying... [id=test-project1-274319, 10s elapsed]
google_compute_shared_vpc_host_project.host: Destruction complete after 11s

Destroy complete! Resources: 2 destroyed.
```

**Comment by pfilourenco** [comment]
-----------
Hello, this is what I have. I also have more resources on this terraform.
----------

 # google_compute_shared_vpc_host_project.host_project will be destroyed
  - resource "google_compute_shared_vpc_host_project" "host_project" {
      - id      = "XXX-network-projectXXX-d-1765" -> null
      - project = "XXX-network-projectXXX-d-1765" -> null
    }

  # google_compute_shared_vpc_service_project.service_project_1 will be destroyed
  - resource "google_compute_shared_vpc_service_project" "service_project_1" {
      - host_project    = "XXX-network-projectXXX-d-1765" -> null
      - id              = "XXX-network-projectXXX-d-1765/ddc-YYY" -> null
      - service_project = "ddc-YYY" -> null
    }

Plan: 0 to add, 0 to change, 28 to destroy.

Do you really want to destroy all resources?
  Terraform will destroy all your managed infrastructure, as shown above.
  There is no undo. Only 'yes' will be accepted to confirm.

  Enter a value: yes


google_compute_shared_vpc_service_project.service_project_1: Destroying... [id=XXX-network-projectXXX-d-1765/ddc-YYY]
google_compute_shared_vpc_service_project.service_project_1: Destruction complete after 2s
google_compute_shared_vpc_host_project.host_project: Destroying... [id=XXX-network-projectXXX-d-1765]


Error: Error disabling Shared VPC Host "XXX-network-projectXXX-d-1765": googleapi: Error 400: Cannot disable project as a shared VPC host because it has active service projects., badRequest
	
---------------------------
When I try again : 
---------------------------

google_compute_shared_vpc_host_project.host_project: Refreshing state... [id=XXX-network-projectXXX-d-1765]
data.google_compute_zones.available: Refreshing state...

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  # google_compute_shared_vpc_host_project.host_project will be destroyed
  - resource "google_compute_shared_vpc_host_project" "host_project" {
      - id      = "XXX-network-projectXXX-d-1765" -> null
      - project = "XXX-network-projectXXX-d-1765" -> null
    }

Plan: 0 to add, 0 to change, 1 to destroy.

Do you really want to destroy all resources?
  Terraform will destroy all your managed infrastructure, as shown above.
  There is no undo. Only 'yes' will be accepted to confirm.

  Enter a value: yes

google_compute_shared_vpc_host_project.host_project: Destroying... [id=XXX-network-projectXXX-d-1765]

Error: Error disabling Shared VPC Host "XXX-network-projectXXX-d-1765": googleapi: Error 400: Cannot disable project as a shared VPC host because it has active service projects., badRequest

**Comment by MikiLoz92** [comment]
I'm also having this issue. Terraform says it has completed destroying a service project:
```
module.github_runners.google_compute_shared_vpc_service_project.service: Destruction complete after 2s
```
But then I still see it on the GCP dashboard:
![image](https://user-images.githubusercontent.com/786375/80953792-1c9fe900-8df4-11ea-8a2c-401d56d68e71.png)
When I try to disable the shared VPC host it's failing in the same fashion as @pfilourenco.

**Comment by pfilourenco** [comment]
Hi,

@venkykuberan @MikiLoz92 

Yes, on GCP GUI I can see the service project still attached:
![image](https://user-images.githubusercontent.com/64209266/80956374-05172f00-8df9-11ea-84ec-39cebf204096.png)

I can detach without problems, via GUI:
![image](https://user-images.githubusercontent.com/64209266/80956418-1f510d00-8df9-11ea-8f0a-eec6b71e4f4e.png)

Then, on Terraform I can destroy successfully.

```
data.google_compute_zones.available: Refreshing state...
google_compute_shared_vpc_host_project.host_project: Refreshing state... [id=nos-network-hybcloudsec-d-1765]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  # google_compute_shared_vpc_host_project.host_project will be destroyed
  - resource "google_compute_shared_vpc_host_project" "host_project" {
      - id      = "nos-network-hybcloudsec-d-1765" -> null
      - project = "nos-network-hybcloudsec-d-1765" -> null
    }

Plan: 0 to add, 0 to change, 1 to destroy.

Do you really want to destroy all resources?
  Terraform will destroy all your managed infrastructure, as shown above.
  There is no undo. Only 'yes' will be accepted to confirm.

  Enter a value: yes

google_compute_shared_vpc_host_project.host_project: Destroying... [id=nos-network-hybcloudsec-d-1765]
google_compute_shared_vpc_host_project.host_project: Still destroying... [id=nos-network-hybcloudsec-d-1765, 10s elapsed]
google_compute_shared_vpc_host_project.host_project: Destruction complete after 12s

Destroy complete! Resources: 1 destroyed.
root@WLT717006:/automation/gcp#
```

Looks like Terraform is not detaching the service project. 

BR,
Pedro Lourenço

**Comment by pdemagny** [comment]
Hi, i appear to be suffering from the same issue too.

I have a 38 resources configuration coming from several modules. But i think the same logic applies.  
I have a host project, with a service project attached to it.  

The two projects and the attachment comes from this module:

```hcl
# Generate the random part ot project's id
resource "random_id" "random_project_id" {
  keepers = {
    project_id = var.general["name"]
  }
  byte_length = 3
}

# Create a regular project
resource "google_project" "gcp_project" {
  name                = var.general["name"]
  project_id          = "${var.general["name"]}-${random_id.random_project_id.hex}"
  org_id              = var.settings["org_id"]
  billing_account     = var.settings["billing_account"]
  auto_create_network = lookup(var.settings, "auto_create_network", true)
  labels              = var.labels
}

# Enable API Services on this project
resource "google_project_service" "enable_api_services" {
  for_each = toset(var.api_services)
  project            = google_project.gcp_project.project_id
  service            = each.key
  disable_on_destroy = false
}

# Enable host project feature for the project, if it's a shared vpc host project
resource "google_compute_shared_vpc_host_project" "shared_vpc_host_project" {
  count      = var.settings["shared_vpc_project_type"] == "host" ? 1 : 0
  depends_on = [google_project_service.enable_api_services]
  project = google_project.gcp_project.project_id
}

# Get project's default service account
data "google_compute_default_service_account" "default_sa" {
  project = google_project_service.enable_api_services["compute.googleapis.com"].project
}

# Allow service accounts to use vpc networking at project level, if project is a service project
resource "google_project_iam_binding" "network_users_project_binding" {
  count      = var.settings["shared_vpc_project_type"] == "service" ? 1 : 0
  depends_on = [google_project_service.enable_api_services]
  project = var.settings["shared_vpc_host_project_id"]
  role    = "roles/compute.networkUser"
  members = [
    local.s_account_fmt,
    local.api_s_account_fmt,
    local.gke_s_account_fmt,
  ]
}

# Allow service accounts to use vpc networking at cluster level, if project is a service project
resource "google_project_iam_binding" "gke_host_agent_project_binding" {
  count      = var.settings["shared_vpc_project_type"] == "service" ? 1 : 0
  depends_on = [google_project_service.enable_api_services]
  project = var.settings["shared_vpc_host_project_id"]
  role    = "roles/container.hostServiceAgentUser"
  members = [
    local.s_account_fmt,
    local.api_s_account_fmt,
    local.gke_s_account_fmt,
  ]
}

# Link service project to host project
resource "google_compute_shared_vpc_service_project" "service_project_attachment" {
  count = var.settings["shared_vpc_project_type"] == "service" ? 1 : 0
  host_project    = var.settings["shared_vpc_host_project_id"]
  service_project = google_project.gcp_project.project_id
}
```

**When i destroy everything, Terraform says that the attachments & the service project are destroyed**:

```bash
module.service_project1.google_compute_shared_vpc_service_project.service_project_attachment[0]: Destroying... [id=tf-dev-host1-9ccb2c/tf-dev-service1-911c79]
module.service_project1.google_compute_shared_vpc_service_project.service_project_attachment[0]: Destruction complete after 1s
module.service_project1.google_project.gcp_project: Destroying... [id=projects/tf-dev-service1-911c79]
module.service_project1.google_project.gcp_project: Destruction complete after 4s
module.host_project1.google_compute_shared_vpc_host_project.shared_vpc_host_project[0]: Destroying... [id=tf-dev-host1-9ccb2c]

Error: Error disabling Shared VPC Host "tf-dev-host1-9ccb2c": googleapi: Error 400: Cannot disable project as a shared VPC host because it has active service projects., badRequest
```

**Yet it isn't true, Terraform didn't remove the attachment between these two project, which i can confirm both in the GCP Dashboard as other have already said, and with the gcloud cli**:

```bash
❯ gcloud compute shared-vpc associated-projects list tf-dev-host1-9ccb2c                                        
RESOURCE_ID             RESOURCE_TYPE
tf-dev-service1-911c79  PROJECT
```

If i manually remove the attachment like this:
```bash
❯ gcloud compute shared-vpc associated-projects remove tf-dev-service1-911c79 --host-project tf-dev-host1-9ccb2c
Updated [https://www.googleapis.com/compute/v1/projects/tf-dev-host1-9ccb2c].
```

Only then i can destroy the host project normally ...

**Comment by ogarza** [comment]
I have the exact same problem on v 0.12.24.

Host project with shared vpc.
Two service projects attached.
```resource "google_compute_shared_vpc_service_project" "archive_shared_vpc_attachment" {
   
    host_project = var.host_vpc_project
    service_project = google_project.archive.project_id
    
    depends_on = [google_project_service.archive_project_services]
}``` is what I am using in the service projects.

When I run terraform destroy I get an error message saying that there are still service projects active

If I run terraform destroy again it works and deletes everything.

Really annoying that I have not figured out how to work around this yet.

**Comment by venkykuberan** [comment]
@pfilourenco  sorry for the delay in responding. We see 2 issues being mentioned here

1) Not able to destroy/disable host project
 2) Although terraform/provider confirms the deletion of the service project, Console still shows the project.

Can you please attach the debug logs to understand what's going on ?

**Comment by pdemagny** [comment]
Hello everyone,

I've been trying out different solutions to workaround this problem. And to me at least, it looks like there no bug but only dependencies issues.

I have a main configuration composed of several modules, and the problem goes away when i set a dummy links between my modules, like so (look for the **wait_modules** keys):

```hcl
module "shared_vpc" {
  source = "../terraform-modules/gcp_shared_vpc"

  general = {
    region = local.workspace.region
  }

  settings = {
    host_project    = module.host_project1.shared_vpc_host_project_id
    service_project = module.service_project1.project_id
    service_accounts = [
      module.service_project1.s_account_fmt,
      module.service_project1.api_s_account_fmt,
      module.service_project1.gke_s_account_fmt,
    ]
  }
}

...

module "bastion_host" {
  source = "../terraform-modules/gcp_compute_instance"

  providers = {
    google      = google
    google-beta = google-beta
  }

  general = {
    project_id   = module.service_project1.project_id
    name         = "${local.workspace.prefix}-bastion1-vm"
    region       = local.workspace.region
    zone         = local.workspace.zone
    wait_modules = [module.shared_vpc.service_project_attachment_id, module.vpc_net1_router.name]
  }

  settings = {
    # removed for clarity
  }
}

module "gke_cluster1" {
  source = "../terraform-modules/gcp_gke_cluster"

  providers = {
    google      = google
    google-beta = google-beta
  }

  general = {
    project      = module.service_project1.project_id
    name         = "${local.workspace.prefix}-gke-cluster1"
    location     = local.workspace.gke_cluster1_location
    wait_modules = [module.shared_vpc.service_project_attachment_id, module.shared_vpc.gke_host_agent_etag, module.vpc_net1_router.name]
  }

  settings = {
    # removed for clarity
  }
}
```

It's like an artificial depends_on for modules that will hopefully be obsolete when Terraform 0.13 comes out !
I hope it will help others to solve this.

**Comment by paulspiegel** [comment]
I don't there's a module dependency issue in my case.  I removed a single module reference that was defining one google_compute_shared_vpc_service_project and two google_compute_subnetwork_iam_member resources.

The apply looked successful:
```
module.sharedvpc_access_cloud.google_compute_subnetwork_iam_member.network_user["maintainer_to_nonprod_east"]: Destroying... [id=***]
module.sharedvpc_access_cloud.google_compute_shared_vpc_service_project.project: Destroying... [id=***]
module.sharedvpc_access_cloud.google_compute_shared_vpc_service_project.project: Destruction complete after 2s
module.sharedvpc_access_cloud.google_compute_subnetwork_iam_member.network_user["maintainer_to_nonprod_central"]: Destroying... [id=***]
```

However both the service project still exists and the network.user membership is also still assigned.  This is Terraform 0.12.24 / Google provider 3.29.0

**Comment by aceresia** [comment]
I'm still experiencing this issue.  Any progress?

**Comment by ScottSuarez** [comment]
Are any users still encountering this issue? If so do you have recent examples you can share. It has been a long time since this issue has been filed and we may close this issue as stale. If it it is still occurring we can notify the appropriate team.

**Comment by rahulracker7539** [comment]
am getting below error while removing service project from the host project through terraform. I have already deleted all the dependencies which were using host prject nw but still getting below error while applying.

Error: Error disabling Shared VPC Resource "service project": Error waiting for Disabling Shared VPC Resource: The resource 'projects/<service project>/global/networkInstances/<randomid>' is still linked to shared VPC host 'projects/<host project>'.

**Comment by MartinNowak** [comment]
> Error: Error disabling Shared VPC Resource "service project": Error waiting for Disabling Shared VPC Resource: The resource 'projects//global/networkInstances/' is still linked to shared VPC host 'projects/'.

<details><summary>Details</summary>
<p>

```json
{
  "protoPayload": {
    "@type": "type.googleapis.com/google.cloud.audit.AuditLog",
    "status": {
      "code": 3,
      "message": "INVALID_USAGE",
      "details": [
        {
          "@type": "type.googleapis.com/google.protobuf.Struct",
          "value": {
            "invalidUsage": {
              "userVisibleReason": "The resource 'projects/gcp-service-project-xxx/global/networkInstances/v-123456789-7e4f9833-c8cf-4ba5-8b9c-027471e150e7' is still linked to shared VPC host 'projects/gcp-host-project-xxx'.",
              "resource": {
                "resourceType": "PROJECT",
                "resourceName": "123xxxxxxxxx",
                "project": {
                  "canonicalProjectId": "123xxxxxxxxx"
                },
                "scope": {
                  "scopeType": "GLOBAL",
                  "scopeName": "global"
                }
              }
            }
          }
        }
      ]
    },
    "authenticationInfo": {
      "principalEmail": "terraform@gcp-tf-project-xxx.iam.gserviceaccount.com",
      "serviceAccountDelegationInfo": [
        {
          "principalSubject": "principal://iam.googleapis.com/projects/xxx"
        }
      ],
      "principalSubject": "serviceAccount:terraform@gcp-tf-project-xxx.iam.gserviceaccount.com"
    },
    "requestMetadata": {
      "callerIp": "40.83.59.87",
      "callerSuppliedUserAgent": "google-api-go-client/0.5 Terraform/1.7.3 (+https://www.terraform.io) Terraform-Plugin-SDK/2.33.0 terraform-provider-google/dev,gzip(gfe)",
      "requestAttributes": {},
      "destinationAttributes": {}
    },
    "serviceName": "compute.googleapis.com",
    "methodName": "v1.compute.projects.disableXpnResource",
    "resourceName": "projects/gcp-host-project-xxx",
    "request": {
      "@type": "type.googleapis.com/compute.projects.disableXpnResource"
    }
  },
  "insertId": "w6rg7ae11ul8",
  "resource": {
    "type": "gce_project",
    "labels": {
      "project_id": "100xxxxxxxxx"
    }
  },
  "timestamp": "2024-08-05T07:21:03.473596Z",
  "severity": "ERROR",
  "labels": {
    "compute.googleapis.com/root_trigger_id": "94d26e70-92a0-48cc-9177-3f5af1145989"
  },
  "logName": "projects/gcp-host-project-xxx/logs/cloudaudit.googleapis.com%2Factivity",
  "operation": {
    "id": "operation-123456789012-xxx",
    "producer": "compute.googleapis.com",
    "last": true
  },
  "receiveTimestamp": "2024-08-05T07:21:03.949657518Z"
}
```

</p>
</details>

**Comment by derhally** [comment]
I ran into this issue today with the following error indicating a VM is still attached when it was deleted 1+ min before attempting to destroy the `google_compute_shared_vpc_service_project` resource to detach the service project. Going to the shared VPC page of "foo-76f0aaeb-app" project in the GCP Console shows no subnets but still says "Host project foo-76f0aaeb-vpc is sharing its subnets with this project."

> Error: Error disabling Shared VPC Resource "foo-76f0aaeb-app": Error waiting for Disabling Shared VPC Resource: The resource 'projects/foo-76f0aaeb-app/zones/us-central1-a/instances/bastion-usc1' is still linked to shared VPC host 'projects/foo-76f0aaeb-vpc'.

**Comment by thanos-kataras** [comment]
TLDR; You can create a `google_compute_shared_vpc_service_project` without a VPC but you can't delete the VPC on the hosting project without first deleting `google_compute_shared_vpc_service_project`

For anyone that may have the same issue:

Problem code:
```
resource "google_compute_shared_vpc_service_project" "host_service_project_attach" {
  host_project    = module.host_project.project_id
  service_project = module.service_project.project_id
}
```
Working code:
```
resource "google_compute_shared_vpc_service_project" "host_service_project_attach" {
  host_project    = module.host_project.project_id
  service_project = module.service_project.project_id

  depends_on = [module.vpc]
}
```
Not the vpc will only be deleted after the attachment itself.
